src/main/java/com/minecolonies/coremod/client/gui/WindowGuiCrafting.java

```diff
@@ -0,0 +1,168 @@
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Crafting gui.
+ */
+public class WindowGuiCrafting extends GuiContainer
+{
+    private static final ResourceLocation CRAFTING_TABLE_GUI_TEXTURES = new ResourceLocation(Constants.MOD_ID, "textures/gui/crafting2x2.png");
+
+    /**
+     * X offset of the button.
+     */
+    private static final int BUTTON_X_OFFSET = 75;
+
+    /**
+     * Y offset of the button.
+     */
+    private static final int BUTTON_Y_POS = 210;
+
+    /**
+     * Button width.
+     */
+    private static final int BUTTON_WIDTH = 150;
+
+    /**
+     * Button height.
+     */
+    private static final int BUTTON_HEIGHT = 20;
+
+    /**
+     * Color of the gui description.
+     */
+    private static final int GUI_COLOR = 4_210_752;
+
+    /**
+     * X offset of the gui description.
+     */
+    private static final int X_OFFSET = 97;
+
+    /**
+     * Y offset of the gui description.
+     */
+    private static final int Y_OFFSET = 8;
+
+    /**
+     * The button to click done after finishing the recipe.
+     */
+    private GuiButton doneButton;
+
+    /**
+     * The building the window belongs to.
+     */
+    private final AbstractBuilding.View building;
+
+    /**
+     * Create a crafting gui window.
+     *
+     * @param playerInv the player.
+     * @param worldIn   the world.
+     * @param building  the building it belongs to.
+     */
+    public WindowGuiCrafting(InventoryPlayer playerInv, World worldIn, final AbstractBuilding.View building)
+    {
+        this(playerInv, worldIn, BlockPos.ORIGIN, building);
+    }
+
+    /**
+     * Create a crafting gui window.
+     *
+     * @param playerInv     the player.
+     * @param worldIn       the world.
+     * @param blockPosition the position.
+     * @param building      the building.
+     */
+    public WindowGuiCrafting(InventoryPlayer playerInv, World worldIn, BlockPos blockPosition, final AbstractBuilding.View building)
+    {
+        super(new CraftingGUIBuilding(playerInv, worldIn, blockPosition));
+        this.building = building;
+    }
+
+    @Override
+    public void initGui()
+    {
+        super.initGui();
+        this.doneButton = this.addButton(new GuiButton(0, this.width / 2 - BUTTON_X_OFFSET, BUTTON_Y_POS, BUTTON_WIDTH, BUTTON_HEIGHT, I18n.format("gui.done", new Object[0])));
+    }
+
+    @Override
+    protected void mouseClicked(final int mouseX, final int mouseY, final int mouseButton) throws IOException
+    {
+        super.mouseClicked(mouseX, mouseY, mouseButton);
+
+        if (doneButton.isMouseOver())
+        {
+            final List<ItemStack> input = new ArrayList<>();
+            final List<ItemStack> secondaryOutput = new ArrayList<>();
+
+            for(int i = 1; i <= 4; i++)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)